kvbc/src/v4blockchain/detail/keys_history.cpp

```diff
@@ -0,0 +1,226 @@
+
+BlockId KeysHistory::last_block_{0};
+
+KeysHistory::KeysHistory(
+    const std::shared_ptr<concord::storage::rocksdb::NativeClient>& native_client,
+    const std::optional<std::map<std::string, concord::kvbc::categorization::CATEGORY_TYPE>>& categories)
+    : native_client_{native_client}, category_mapping_(native_client, categories) {
+  if (native_client_->createColumnFamilyIfNotExisting(v4blockchain::detail::KEYS_HISTORY_CF,
+                                                      KHCompactionFilter::getFilter())) {
+    LOG_INFO(V4_BLOCK_LOG, "Created [" << v4blockchain::detail::KEYS_HISTORY_CF << "] column family for keys history");
+  }
+}
+
+void KeysHistory::addBlockKeys(const concord::kvbc::categorization::Updates& updates,
+                               const BlockId block_id,
+                               storage::rocksdb::NativeWriteBatch& write_batch) {
+  LOG_DEBUG(V4_BLOCK_LOG, "Adding keys of block [" << block_id << "] to the keys history CF");
+
+  last_block_ = block_id;
+  for (const auto& [category_id, category_updates] : updates.categoryUpdates().kv) {
+    std::visit(overloaded{[category_id = category_id, &write_batch, &block_id, this](const auto& updates) {
+                            std::vector<std::string> updates_keys(updates.kv.size());
+                            for (const auto& [k, _] : updates.kv) {
+                              (void)_;  // unused
+                              updates_keys.push_back(k);
+                            }
+                            for (const auto& k : updates.deletes) updates_keys.push_back(k);
+                            handleUpdatesImp(block_id, category_id, updates_keys, write_batch);
+                          },
+                          [category_id = category_id, &write_batch, &block_id, this](
+                              const categorization::ImmutableInput& updates) {
+                            std::vector<std::string> updates_keys(updates.kv.size());
+                            for (const auto& [k, _] : updates.kv) {
+                              (void)_;  // unused
+                              updates_keys.push_back(k);
+                            }
+                            handleUpdatesImp(block_id, category_id, updates_keys, write_batch);
+                          }},
+               category_updates);
+  }
+}
+
+void KeysHistory::handleUpdatesImp(const BlockId block_id,
+                                   const std::string& category_id,
+                                   const std::vector<std::string>& updates_keys,
+                                   concord::storage::rocksdb::NativeWriteBatch& write_batch) {
+  std::string block_key = v4blockchain::detail::Blockchain::generateKey(block_id);
+  ConcordAssertEQ(block_key.size(), sizeof(BlockId));
+  const auto& prefix = category_mapping_.categoryPrefix(category_id);
+  const std::string empty_val = "";
+  for (const auto& k : updates_keys) {
+    auto cf_key = prefix;
```

Comment:
You can use getSliceArray as done in the latst_keys to reduce copies

Reply:
Thanks, done.
I moved getSliceArray definition to blockchain.h, let me know if you think of better place.